proposals/webidl-bindings/Explainer.md

```diff
@@ -0,0 +1,623 @@
+
+For example, building on the previous section's example, a Web IDL value of the
+`$Contact` dictionary type could be produced from three WebAssembly `i32`s by
+the following (strawman text-format) binding expression:
+
+```
+(dict $Contact (utf8-str DOMString off-idx=0 len-idx=1) (as type=long idx=2))
+```
+
+This binding expression can be visualized:
+
+![Binding Expression Diagram](example-binding-expression.svg)
+
+
+## Binding Maps
+
+A Web IDL **binding map** allows mapping between *entire tuples* of WebAssembly
+and Web IDL values via a *tuple* of binding expressions.  All binding
+expressions consume the same source tuple.  In contrast, the destination tuple
+of the binding map is created by concatenating the destination tuples of each
+individual binding expression.
+
+Binding maps are either incoming or outgoing.  **Incoming binding maps** contain
+only incoming binding expressions and are applied to the parameters of exported
+functions and the results of imported functions.  Conversely, **outgoing binding
+maps** contain only outgoing binding expressions and are applied to the results of
+exported functions or the parameters of imported functions.
+
+For example, building on the previous example, to create the *parameter* tuple to pass to this Web IDL method:
+
+```
+interface ContactDB {
+  bool addContact(Contact contact, DOMString contactBook);
+};
+```
+
+the following (strawman text-format, which is just a sequence of binding
+expressions) binding map could be used:
+
+```
+(dict $Contact (utf8-str DOMString off-idx=0 len-idx=1) (as type=long idx=2))
+(utf8-str DOMString off-idx=3 len-idx=4)
+```
+
+This binding map can be visualized:
+
+![Binding Map Diagram](example-binding-map.svg)
+
+When the binding map is compiled to machine code by an engine implementing Web
+IDL Bindings, the stub can map the `i32` WebAssembly values to the positional,
+bounds-checked pointer+length arguments that are ideal for the underlying native
+host implementation of `addContact`:
+
+![Binding Map Stub Diagram](example-binding-map-stub.svg)
+
+
+## Function Bindings
+
+To bind an *entire function*, there are five ingredients:
+* a bit indicating "import" or "export"
+* a WebAssembly [`functype`]  (defined by the WebAssembly [Type Section])
+* a Web IDL Function type (defined in the Web IDL [Type Subsection](#web-idl-types))
+* an (outgoing for import, incoming for export) binding map for the parameters
+* an (incoming for import, outgoing for export) binding map for the results
+
+For example, building on the previous example, the following (strawman
+text-format) statement would define a Web IDL Function type compatible with
+`addContact`:
+
+```
+(webidl-type $AddContactFuncWebIDL (func (method any) (param $Contact DOMString) (result bool)))
```

Comment:
The syntax here is a little confusing since `(param $foo i32)` is currently a definition of a single i32 parameter named $foo, but it seems this is actually defining two parameters `$Contact` and `DOMString`.

Reply:
Ah good point; I'll wrap `$Contact` in a `(dict $Contact)`.